snakemake/logging.py

```diff
@@ -305,12 +243,504 @@ def log_handler(self, msg):
+        output = [f"Error in rule {msg['name']}:"]
+
+        if msg["msg"]:
+            output.append(f"    message: {msg['msg']}")
+        output.append(f"    jobid: {msg['jobid']}")
+        if msg["input"]:
+            output.append(f"    input: {', '.join(msg['input'])}")
+        if msg["output"]:
+            output.append(f"    output: {', '.join(msg['output'])}")
+        if msg["log"]:
+            output.append(
+                f"    log: {', '.join(msg['log'])} (check log file(s) for error details)"
+            )
+        if msg["conda_env"]:
+            output.append(f"    conda-env: {msg['conda_env']}")
+        if msg["shellcmd"]:
+            output.append(
+                f"    shell:\n        {msg['shellcmd']}\n        (command exited with non-zero exit code)"
+            )
+
+        for item in msg["aux"].items():
+            output.append(f"    {item[0]}: {item[1]}")
+
+        if self.show_failed_logs and msg["log"]:
+            output.extend(show_logs(msg["log"]))
+
+        return output
+
+    def format_group_error(self, msg):
+        """Helper method to format group error details."""
+        output = [f"Error in group {msg['groupid']}:"]
+
+        if msg["msg"]:
+            output.append(f"    message: {msg['msg']}")
+        if msg["aux_logs"]:
+            output.append(
+                f"    log: {', '.join(msg['aux_logs'])} (check log file(s) for error details)"
+            )
+
+        output.append("    jobs:")
+        for info in msg["job_error_info"]:
+            output.append(f"        rule {info['name']}:")
+            output.append(f"            jobid: {info['jobid']}")
+            if info["output"]:
+                output.append(f"            output: {', '.join(info['output'])}")
+            if info["log"]:
+                output.append(
+                    f"            log: {', '.join(info['log'])} (check log file(s) for error details)"
+                )
+
+        logs = msg["aux_logs"] + [
+            f for info in msg["job_error_info"] for f in info["log"]
+        ]
+        if self.show_failed_logs and logs:
+            output.extend(self.show_logs(logs))
+
+        return output
+
+    def timestamp(self):
+        """Helper method to format the timestamp."""
+        return f"[{time.asctime()}]"
+
+    def format_percentage(self, done, total):
+        """Helper method to format percentage."""
+        if done == total:
+            return "100%"
+        if done == 0:
+            return "0%"
+        precision = 0
+        fraction = done / total
+        fmt_precision = "{{:.{}%}}".format
+        fmt = lambda fraction: fmt_precision(precision).format(fraction)
+        while fmt(fraction) == "100%" or fmt(fraction) == "0%":
+            precision += 1
+        return fmt(fraction)
+
+
+def show_logs(logs):
+    """Helper method to show logs."""
+    for f in logs:
+        try:
+            content = open(f, "r").read()
```

Comment:
I always thought the idiomatic version of this uses a with statement, even if one reads the entire file into memory.